mne/realtime/base_client.py

```diff
@@ -0,0 +1,196 @@
+# Authors: Teon Brooks <teon.brooks@gmail.com>
+#          Mainak Jas <mainak@neuro.hut.fi>
+#
+# License: BSD (3-clause)
+
+import threading
+import time
+import numpy as np
+
+from ..utils import logger, fill_doc
+
+def _buffer_recv_worker(client):
+    """Worker thread that constantly receives buffers."""
+    try:
+        for raw_buffer in client.iter_raw_buffers():
+            client._push_raw_buffer(raw_buffer)
+    except RuntimeError as err:
+        # something is wrong, the server stopped (or something)
+        client._recv_thread = None
+        print('Buffer receive thread stopped: %s' % err)
+
+
+class _BaseClient(object):
+    """Base Realtime Client.
+
+    Parameters
+    ----------
+    identifier : str
+        The identifier of the server. IP address or LSL id or raw filename.
+    port : int | None
+        Port to use for the connection.
+    wait_max : float
+        Maximum time (in seconds) to wait for real-time buffer to start.
+    tmin : float | None
+        Time instant to start receiving buffers. If None, start from the latest
+        samples available.
+    tmax : float
+        Time instant to stop receiving buffers.
+    buffer_size : int
+        Size of each buffer in terms of number of samples.
+    verbose : bool, str, int, or None
+        If not None, override default verbose level (see :func:`mne.verbose`
+        and :ref:`Logging documentation <tut_logging>` for more).
+    """
+
+    def __init__(self, identifier, port=None, wait_max=10., tmin=None,
+                 tmax=np.inf, buffer_size=1000, verbose=None):  # noqa: D102
+        self.identifier = identifier
+        self.port = port
+        self.wait_max = wait_max
+        self.tmin = tmin
+        self.tmax = tmax
+        self.buffer_size = buffer_size
+        self.verbose = verbose
+
+    def __enter__(self):  # noqa: D105
+
+        # connect to buffer
+        logger.info("Client: Waiting for server to start")
+        start_time, current_time = time.time(), time.time()
+        success = False
+        while current_time < (start_time + self.wait_max):
+            try:
+                self.connect()
+                logger.info("Client: Connected")
+                success = True
+                break
+            except Exception:
+                current_time = time.time()
+                time.sleep(0.1)
+
+        if not success:
+            raise RuntimeError('Could not connect to Buffer')
+
+        self.create_info()
+        self._enter_extra()
+
+        return self
```

Comment:
I was bothered by the non-nicety of the explicit flag and wanted to make a PR:

```
        logger.info("Client: Waiting for server to start")
        start_time, current_time = time.time(), time.time()
        while current_time < (start_time + self.wait_max):
            try:
                self.connect()
                logger.info("Client: Connected")
                self.create_info()
                self._enter_extra()

                return self
            except Exception:
                current_time = time.time()
                time.sleep(0.1)
        # if no success before the while exits ...
        raise RuntimeError('Could not connect to Buffer')
```

But I somehow can't PR onto this PR?

Anyways, wdyt?

Reply:
wait, i don't see the difference other than the upon success now, it won't create the info or the call the `_enter_extra`? to make a pr on my pr, make sure you are trying to attach it to my fork head with the `rtclient` branch

Reply:
I simply didn't like the `success` flag :) Functionally it's *hopefully* the same.

Reply:
Personally I would make use of the under-utilized `else` condition of loops (meaning, "if did not break"), since it explicitly does what your comment was trying to indicate:
```
        logger.info("Client: Waiting for server to start")
        start_time, current_time = time.time(), time.time()
        while current_time < (start_time + self.wait_max):
            try:
                self.connect()
                logger.info("Client: Connected")
                self.create_info()
                self._enter_extra()
                break
            except Exception:
                current_time = time.time()
                time.sleep(0.1)
        else:
            raise RuntimeError('Could not connect to Buffer')
        return self
```

Reply:
Oh the else triggers if it doesn’t break? If yes that’s exactly what I wanted.

Reply:
ah, gotcha, i see